x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -102,50 +102,62 @@ public AsyncSender interceptSender(AsyncSender sender) {
+                // ourselves otherwise we wind up using a version newer than what we can actually send
+                final Version minVersion = Version.min(connection.getVersion(), Version.CURRENT);
+
+                // Sometimes a system action gets executed like a internal create index request or update mappings request
+                // which means that the user is copied over to system actions so we need to change the user
+                if (AuthorizationUtils.shouldReplaceUserWithSystem(threadPool.getThreadContext(), action)) {
+                    securityContext.executeAsUser(SystemUser.INSTANCE, (original) -> sendWithUser(connection, action, request, options,
+                            new ContextRestoreResponseHandler<>(threadPool.getThreadContext().wrapRestorable(original)
+                                    , handler), sender, requireAuth), minVersion);
+                } else if (AuthorizationUtils.shouldSetUserBasedOnActionOrigin(threadPool.getThreadContext())) {
+                    AuthorizationUtils.switchUserBasedOnActionOriginAndExecute(threadPool.getThreadContext(), securityContext,
+                            (original) -> sendWithUser(connection, action, request, options,
+                                    new ContextRestoreResponseHandler<>(threadPool.getThreadContext().wrapRestorable(original)
+                                            , handler), sender, requireAuth));
+                } else if (securityContext.getAuthentication() != null &&
+                        securityContext.getAuthentication().getVersion().equals(minVersion) == false) {
+                    // re-write the authentication since we want the authentication version to match the version of the connection
+                    securityContext.executeAfterRewritingAuthentication(original -> sendWithUser(connection, action, request, options,
+                        new ContextRestoreResponseHandler<>(threadPool.getThreadContext().wrapRestorable(original), handler), sender,
+                        requireAuth), minVersion);
                 } else {
-                    sender.sendRequest(connection, action, request, options, handler);
+                    sendWithUser(connection, action, request, options, handler, sender, requireAuth);
                 }
             }
         };
     }
 
+    /**
+     * Based on the current cluster state &amp; license, should we require that all outgoing actions have an authentication header
+     * of some sort?
+     */
+    private boolean shouldRequireExistingAuthentication() {
+        // make a local copy of isStateNotRecovered as this is a volatile variable and it
+        // is used multiple times in the method. The copy to a local variable allows us to
+        // guarantee we use the same value wherever we would check the value for the state
+        // being recovered
+        final boolean stateNotRecovered = isStateNotRecovered;
```

Comment:
do we still need to copy to a local variable that we use only once directly below ?